File: lib/python-sdk/common_grants_sdk/extensions/transforms.py
"""build_transforms() — generates to_common/from_common callables from mapping dicts.

Using this utility is optional — plugin authors may provide plain hand-written
callables instead.

Mappings are validated at call time. Custom handler names are
registered per call only; name collisions with defaults raise at call time
rather than silently shadowing them.
"""

from __future__ import annotations

from typing import Any, Callable, TypeVar, overload

from pydantic import BaseModel, ValidationError

from common_grants_sdk.utils.transformation import (
    DEFAULT_HANDLERS,
    HandlerError,
    transform_from_mapping,
)

from .types import Handler, TransformError, TransformResult

TCommon = TypeVar("TCommon", bound=BaseModel)
TSource = TypeVar("TSource", bound=BaseModel)


def _validate_output_paths(
    mapping: dict[str, Any],
    model: type[BaseModel],
    known_handlers: set[str],
    direction: str = "to_common",
) -> None:
    """Validate that top-level output keys in mapping are valid fields on model.

    Called when common_schema is supplied to build_transforms(). Custom fields
    declared by the plugin appear as regular model fields on the generated common
    model and are therefore treated as valid output paths automatically.

    Raises ValueError if any top-level key is not a field name or alias on model.
    """
    valid_names: set[str] = set(model.model_fields.keys())
    for field_info in model.model_fields.values():
        if field_info.alias:
            valid_names.add(field_info.alias)

    # Top-level handler invocations (rare but structurally valid) are not output keys
    output_keys = {k for k in mapping if k not in known_handlers}
    invalid = output_keys - valid_names
    if invalid:
        noun = "field" if len(invalid) == 1 else "fields"
        raise ValueError(
            f"build_transforms ({direction}_mapping): unknown output {noun} "
            f"{sorted(invalid)!r} for model {model.__name__}. "
            f"Declare them as custom_fields in SchemaInput or check the field name."
        )


def _validate_mapping(mapping: Any, known_handlers: set[str], path: str = "") -> None:
    """Walk the mapping tree and raise ValueError on structural malformation.

    For each dict node:
    - If a key is a known handler, the node must contain ONLY that handler key.
      The corresponding value is a runtime-only handler argument and is NOT
      recursed into.
    - All other keys are output field names (always valid); their values are
      recursed into.

    Raises ValueError if any node is not a dict, string, number, boolean, or None
    (e.g. a list where a scalar or dict is expected), or if a handler key appears
    alongside sibling keys in the same dict (ambiguous — handler invocations must
    be the sole key in their dict).

    Note: this function cannot detect intended-but-unknown handler invocations
    because unknown keys are indistinguishable from output field names at static
    analysis time. That detection is deferred to the full SDK.
    """
    if mapping is None or isinstance(mapping, (str, int, float, bool)):
        return  # primitives and None are valid literals

    if not isinstance(mapping, dict):
        raise ValueError(
            f"Invalid mapping node at '{path}': expected dict, str, number, or bool, "
            f"got {type(mapping).__name__}"
        )

    handler_keys = [k for k in mapping if k in known_handlers]
    if handler_keys and len(mapping) > 1:
        label = f" at '{path}'" if path else ""
        raise ValueError(
            f"Invalid mapping node{label}: handler key {handler_keys[0]!r} "
            f"cannot have sibling keys {sorted(k for k in mapping if k not in known_handlers)!r}. "
            f"A handler invocation must be the only key in its dict."
        )

    for key, value in mapping.items():
        current_path = f"{path}.{key}" if path else key
        if key in known_handlers:
            # Handler invocation — argument is runtime-only, do not recurse
            continue
        _validate_mapping(value, known_handlers, current_path)


@overload
def build_transforms(
    to_common_mapping: dict[str, Any],
    from_common_mapping: dict[str, Any],
    handlers: dict[str, Handler] | None = ...,
    common_schema: None = ...,
    source_schema: None = ...,
) -> tuple[
    Callable[[Any], TransformResult[Any]],
    Callable[[Any], TransformResult[Any]],
]: ...


@overload
def build_transforms(
    to_common_mapping: dict[str, Any],
    from_common_mapping: dict[str, Any],
    handlers: dict[str, Handler] | None = ...,
    common_schema: type[TCommon] = ...,
    source_schema: None = ...,
) -> tuple[
    Callable[[Any], TransformResult[TCommon | dict[str, Any]]],
    Callable[[Any], TransformResult[Any]],
]: ...


@overload
def build_transforms(
    to_common_mapping: dict[str, Any],
    from_common_mapping: dict[str, Any],
    handlers: dict[str, Handler] | None = ...,
    common_schema: None = ...,
    source_schema: type[TSource] = ...,
) -> tuple[
    Callable[[Any], TransformResult[Any]],
    Callable[[Any], TransformResult[TSource | dict[str, Any]]],
]: ...


@overload
def build_transforms(
    to_common_mapping: dict[str, Any],
    from_common_mapping: dict[str, Any],
    handlers: dict[str, Handler] | None = ...,
    common_schema: type[TCommon] = ...,
    source_schema: type[TSource] = ...,
) -> tuple[
    Callable[[Any], TransformResult[TCommon | dict[str, Any]]],
    Callable[[Any], TransformResult[TSource | dict[str, Any]]],
]: ...


def build_transforms(
    to_common_mapping: dict[str, Any],
    from_common_mapping: dict[str, Any],
    handlers: dict[str, Handler] | None = None,
    common_schema: type[BaseModel] | None = None,
    source_schema: type[BaseModel] | None = None,
) -> tuple[
    Callable[[Any], TransformResult[Any]],
    Callable[[Any], TransformResult[Any]],
]:
    """Generate to_common and from_common callables from mapping dicts.

    Args:
        to_common_mapping: mapping from source system → CommonGrants.
        from_common_mapping: mapping from CommonGrants → source system.
        handlers: Optional additional handlers registered for this call only.
            Keys must not collide with DEFAULT_HANDLERS (raises ValueError if they do).
            Each ``build_transforms()`` call gets its own isolated handler registry —
            handlers passed here are not visible to any other call.

            Example::

                def handle_upper(data, path):
                    parts = path.split(".")
                    val = data
                    for p in parts:
                        val = val.get(p) if isinstance(val, dict) else None
                    return str(val).upper() if val is not None else None

                to_common, _ = build_transforms(
                    {"title": {"upper": "data.opportunity_title"}},
                    {},
                    handlers={"upper": handle_upper},
                )
        common_schema: Optional Pydantic model class to validate the to_common output
            against. Must be the fully extended generated model class (e.g. the
            generated Opportunity from generated/schemas.py), NOT the base class
            (e.g. OpportunityBase). Passing a base class will silently weaken
            validation — custom_fields will only be checked against
            dict[str, CustomField] rather than the typed container produced by the
            plugin's custom field declarations. When provided, model_validate is
            called on the transform result and any ValidationErrors are appended to
            TransformResult.errors rather than raised.

            Note on result shape: when common_schema is set, TransformResult.result
            holds the validated Pydantic instance on success, or the raw transformed
            dict on ValidationError (so callers can inspect the malformed data
            alongside the errors). This is intentional — check TransformResult.errors
            before consuming TransformResult.result.
        source_schema: Optional Pydantic model class to validate the from_common output
            against. Without this, from_common casts its result to a plain dict with
            no runtime check, so the return type provides no real safety guarantee.
            When provided, model_validate is called on the transform result and any
            ValidationErrors are appended to TransformResult.errors rather than raised.
            The result shape follows the same convention as common_schema: a validated
            model instance on success, or the raw dict alongside errors on failure.

    Returns:
        A (to_common, from_common) tuple. Each callable accepts a dict and returns
        TransformResult[Any]. Failures surface as TransformError entries in
        TransformResult.errors rather than being raised.

    Raises:
        ValueError: At call time if handler names collide with defaults,
            or if either mapping has structural malformation.

    TODO (full SDK):
        - Validate field-path resolvability at call time (requires sample data or
          schema introspection).
    """
    # Custom handler names must not shadow defaults
    if handlers:
        collisions = set(handlers) & set(DEFAULT_HANDLERS)
        if collisions:
            raise ValueError(
                f"build_transforms: handler names collide with defaults: {sorted(collisions)}"
            )

    merged = {**DEFAULT_HANDLERS, **(handlers or {})}
    known = set(merged)

    # Validate mapping structure at call time
    _validate_mapping(to_common_mapping, known)
    _validate_mapping(from_common_mapping, known)

    # When common_schema is provided, validate that to_common output keys are real fields
    if common_schema is not None:
        _validate_output_paths(to_common_mapping, common_schema, known, "to_common")
    # When source_schema is provided, validate that from_common output keys are real fields
    if source_schema is not None:
        _validate_output_paths(from_common_mapping, source_schema, known, "from_common")

    def to_common(native: Any) -> TransformResult[Any]:
        try:
            result = transform_from_mapping(native, to_common_mapping, handlers=merged)
        except HandlerError as exc:
            error = TransformError(
                str(exc.cause),
                path=None,
                handler=exc.handler,
                source_value=native,
                cause=exc.cause,
            )
            return TransformResult(result={}, errors=[error])
        except Exception as exc:
            error = TransformError(str(exc), path=None, source_value=native, cause=exc)
            return TransformResult(result={}, errors=[error])

        if common_schema is None:
            return TransformResult(result=result, errors=[])

        try:
            validated = common_schema.model_validate(result)
            return TransformResult(result=validated, errors=[])
        except ValidationError as exc:
            errors = [
                TransformError(
                    e["msg"],
                    path=".".join(str(loc) for loc in e["loc"]),
                )
                for e in exc.errors()
            ]
            return TransformResult(result=result, errors=errors)
        except Exception as exc:
            error = TransformError(str(exc), path=None, source_value=result, cause=exc)
            return TransformResult(result=result, errors=[error])

    def from_common(common: Any) -> TransformResult[Any]:
        try:
            result = transform_from_mapping(
                common, from_common_mapping, handlers=merged
            )
        except HandlerError as exc:
            error = TransformError(
                str(exc.cause),
                path=None,
                handler=exc.handler,
                source_value=common,
                cause=exc.cause,
            )
            return TransformResult(result={}, errors=[error])
        except Exception as exc:
            error = TransformError(str(exc), path=None, source_value=common, cause=exc)
            return TransformResult(result={}, errors=[error])

        if source_schema is None:
            return TransformResult(result=result, errors=[])

        try:
            validated = source_schema.model_validate(result)
            return TransformResult(result=validated, errors=[])
        except ValidationError as exc:
            errors = [
                TransformError(
                    e["msg"],
                    path=".".join(str(loc) for loc in e["loc"]),
                )
                for e in exc.errors()
            ]
            return TransformResult(result=result, errors=errors)
        except Exception as exc:
            error = TransformError(str(exc), path=None, source_value=result, cause=exc)
            return TransformResult(result=result, errors=[error])

    return to_common, from_common

File: lib/ts-sdk/src/extensions/transforms.ts
/**
 * `buildTransforms()` — compile a pair of declarative mapping objects into typed
 * `(toCommon, fromCommon)` callables.
 *
 * Each direction is author-provided — this utility never inverts one into the
 * other, because many-to-one handlers like `switch` are not reversible.
 *
 * @module @common-grants/sdk/extensions
 */

import { z } from "zod";

import { TransformError, type Handler, type TransformResult } from "./types";
import { DEFAULT_HANDLERS, HandlerError, transformFromMapping } from "../utils/transformation";

// ############################################################################
// Internal - mapping structure validation
// ############################################################################

/**
 * Walk the mapping tree and throw on structural malformation.
 *
 * Each node must be a primitive (`string` / `number` / `boolean` / `null` /
 * `undefined`) or a plain object. Arrays and class instances are rejected.
 *
 * Handler arguments are runtime-only and skipped — they may legitimately be
 * arrays, deeply nested specs, or anything else the handler accepts. The
 * walker can't detect *unknown* handler names at static analysis time (an
 * unknown key is indistinguishable from an output field name); detection of
 * unknown top-level output keys is now handled by `validateOutputPaths` when
 * `commonModel` is provided; detection at arbitrary nesting depth remains deferred.
 *
 * Sibling keys at a handler-dispatch node are rejected here. The runtime
 * walker is first-key-wins, so `{ field: "x", const: "fallback" }` would
 * silently drop `const` — almost always an author typo — so fail loud at
 * build time instead. The low-level `transformFromMapping` walker stays
 * lenient so programmatic callers composing partial mappings aren't forced
 * into the strict shape.
 *
 * @internal
 */
function validateMapping(mapping: unknown, knownHandlers: Set<string>, path = ""): void {
  if (mapping === null || mapping === undefined) return;
  const t = typeof mapping;
  if (t === "string" || t === "number" || t === "boolean") return;

  if (t !== "object" || Array.isArray(mapping)) {
    throw new Error(
      `Invalid mapping node at '${path}': expected object, string, number, boolean, or null, got ${
        Array.isArray(mapping) ? "array" : t
      }`
    );
  }

  // A handler invocation must be the sole key in its node — see the function
  // docstring above for the first-key-wins rationale.
  const nodeKeys = Object.keys(mapping as Record<string, unknown>);
  const handlerKeys = nodeKeys.filter(k => knownHandlers.has(k));
  if (handlerKeys.length > 0 && nodeKeys.length > 1) {
    const siblings = nodeKeys.filter(k => !knownHandlers.has(k)).sort();
    throw new Error(
      `Invalid mapping node at '${path === "" ? "<root>" : path}': handler key '${
        handlerKeys[0]
      }' cannot have sibling keys ${JSON.stringify(siblings)}. ` +
        `A handler invocation must be the only key in its dict.`
    );
  }

  for (const [key, value] of Object.entries(mapping as Record<string, unknown>)) {
    const childPath = path === "" ? key : `${path}.${key}`;
    if (knownHandlers.has(key)) {
      // Handler invocation — argument is runtime-only, do not recurse.
      continue;
    }
    validateMapping(value, knownHandlers, childPath);
  }
}

/**
 * Validate that every top-level output key in `mapping` that is not a known
 * handler name is a real field on `commonModel`.
 *
 * Only runs when `commonModel` is an instance of `z.ZodObject` — when it is
 * not (e.g. `ZodRecord`, `ZodUnion`), returns without error. In practice all
 * schemas produced by `withCustomFields()` and `OpportunityBaseSchema` are
 * `ZodObject`s (`.extend()` / `.merge()` preserve the concrete type), so the
 * fallback is a safety net, not an expected code path.
 *
 * @internal
 */
function validateOutputPaths(
  mapping: Record<string, unknown>,
  knownHandlers: Set<string>,
  // `any` lets this accept Zod schemas whose input type differs from their
  // output type (e.g. schemas that use .transform()). `unknown` would reject
  // those valid schemas here.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  commonSchema: z.ZodType<unknown, any>
): void {
  if (!(commonSchema instanceof z.ZodObject)) return;
  const validNames = new Set(Object.keys(commonSchema.shape));
  const outputKeys = Object.keys(mapping).filter(k => !knownHandlers.has(k));
  const invalid = outputKeys.filter(k => !validNames.has(k));
  if (invalid.length === 0) return;
  throw new Error(
    `buildTransforms (toCommonMapping): unknown output fields ${JSON.stringify(invalid.sort())} for schema. ` +
      `If these are custom fields, map them under the top-level "customFields" key ` +
      `(e.g. { customFields: { yourField: ... } }), not as top-level keys, and declare them in schemas[Object].customFields. ` +
      `Otherwise check the field name.`
  );
}

// ############################################################################
// Public - buildTransforms
// ############################################################################

/**
 * Return shape of {@link buildTransforms}.
 */
export interface BuiltTransforms<TSource, TCommon> {
  toCommon: (source: TSource) => TransformResult<TCommon>;
  fromCommon: (common: TCommon) => TransformResult<TSource>;
}

/**
 * Compile a pair of declarative mapping objects into typed
 * `(toCommon, fromCommon)` callables.
 *
 * @internal Plugin authors should use `definePlugin({ schemas: { [Name]: { mappings } } })`
 * instead of calling this directly. `definePlugin()` invokes `buildTransforms()`
 * automatically from the `mappings` entry and wraps the result with schema validation.
 * This function remains exported for unit testing and advanced use cases.
 *
 * @example
 * ```ts
 * const { toCommon, fromCommon } = buildTransforms(
 *   {
 *     id:    { field: "data.opportunity_uuid" },
 *     title: { field: "data.opportunity_title" },
 *     status: {
 *       value: {
 *         match: {
 *           field: "data.opportunity_status",
 *           case: { posted: "open", archived: "closed" },
 *           default: "custom",
 *         },
 *       },
 *     },
 *   },
 *   { data: { opportunity_title: { field: "title" } } },
 * );
 *
 * const result = toCommon(sourceData);
 * if (result.errors.length === 0) use(result.result);
 * ```
 *
 * @remarks
 * Error aggregation is asymmetric across the two failure modes:
 *
 * - Handler failures (a registered handler throws): the mapping walk
 *   short-circuits on the first failure, so `errors` carries exactly one
 *   `TransformError` even when several fields would have failed.
 * - Zod-validation failures (`commonSchema` or `sourceSchema` provided): every
 *   `ZodIssue` is flattened into a separate `TransformError`, so `errors` carries
 *   the full set.
 *
 * Callers writing strict-mode handling should treat any non-empty `errors`
 * as failure regardless of length.
 *
 * @param toCommonMapping - Declarative mapping from source system format → CommonGrants.
 * @param fromCommonMapping - Declarative mapping from CommonGrants → source system format.
 * @param handlers - Optional custom handlers registered for this call only.
 *   Name collisions with {@link DEFAULT_HANDLERS} raise a `TypeError` at call
 *   time rather than silently shadowing the default.
 * @param commonSchema - Optional Zod schema to validate `toCommon` output against.
 *   Must be the fully extended schema (e.g. result of `withCustomFields(...)`) —
 *   not the base schema. Passing the base schema silently weakens validation of
 *   typed custom fields. When provided, `safeParse()` runs on the transform
 *   result and Zod issues are flattened into `TransformResult.errors`.
 *   `TransformError.path` for Zod-flattened issues uses dot notation including
 *   numeric indices (e.g. `"customFields.items.0.value"`).
 * @param sourceSchema - Optional Zod schema to validate `fromCommon` output
 *   against. Without this, `fromCommon` casts its result to `TSource` without
 *   any runtime check, so `TSource` provides no real safety guarantee. When
 *   provided, `safeParse()` runs on the transform result and Zod issues are
 *   flattened into `TransformResult.errors` using the same format as
 *   `commonSchema`.
 *
 * @throws TypeError when custom handler names collide with built-in defaults.
 * @throws Error when either mapping is structurally malformed (sibling keys
 *   on a handler-dispatch node).
 */
export function buildTransforms<TSource = unknown, TCommon = unknown>(
  toCommonMapping: Record<string, unknown>,
  fromCommonMapping: Record<string, unknown>,
  handlers?: Map<string, Handler>,
  // `any` lets callers pass Zod schemas that transform their input into TCommon
  // (e.g. schemas using .transform()). `unknown` would reject those valid schemas.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  commonSchema?: z.ZodType<TCommon, any>,
  // `any` lets callers pass Zod schemas that transform their input into TSource
  // (e.g. schemas using .transform()). `unknown` would reject those valid schemas.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  sourceSchema?: z.ZodType<TSource, any>
): BuiltTransforms<TSource, TCommon> {
  if (handlers) {
    const collisions = [...handlers.keys()].filter(k => DEFAULT_HANDLERS.has(k));
    if (collisions.length > 0) {
      throw new TypeError(
        `buildTransforms: handler names collide with defaults: ${JSON.stringify(collisions.sort())}`
      );
    }
  }

  const merged = new Map([...DEFAULT_HANDLERS, ...(handlers ?? [])]);
  const known = new Set(merged.keys());

  // Validate mapping structure up front so structurally malformed mappings
  // fail at build time, not on first invocation.
  validateMapping(toCommonMapping, known);
  validateMapping(fromCommonMapping, known);
  if (commonSchema !== undefined) {
    validateOutputPaths(toCommonMapping, known, commonSchema);
  }
  if (sourceSchema !== undefined) {
    validateOutputPaths(fromCommonMapping, known, sourceSchema);
  }

  const runMapping = (
    data: unknown,
    mapping: Record<string, unknown>
  ): { ok: true; value: unknown } | { ok: false; error: TransformError } => {
    try {
      return { ok: true, value: transformFromMapping(data, mapping, { handlers: merged }) };
    } catch (exc) {
      if (exc instanceof HandlerError) {
        const cause = exc.cause;
        return {
          ok: false,
          error: new TransformError(cause instanceof Error ? cause.message : String(cause), {
            handler: exc.handler,
            sourceValue: data,
            cause,
          }),
        };
      }
      return {
        ok: false,
        error: new TransformError(exc instanceof Error ? exc.message : String(exc), {
          sourceValue: data,
          cause: exc,
        }),
      };
    }
  };

  const toCommon = (source: TSource): TransformResult<TCommon> => {
    const ran = runMapping(source, toCommonMapping);
    if (!ran.ok) return { result: {} as TCommon, errors: [ran.error] };

    if (commonSchema === undefined) {
      return { result: ran.value as TCommon, errors: [] };
    }

    const parsed = commonSchema.safeParse(ran.value);
    if (parsed.success) {
      return { result: parsed.data, errors: [] };
    }
    const errors = parsed.error.issues.map(issue => {
      // Root-level issues (e.g. from `.refine()` on the schema itself) have an
      // empty `path` — leave TransformError.path undefined so the "if known" contract
      // in the docstring holds.
      const joined = issue.path.length > 0 ? issue.path.map(p => String(p)).join(".") : undefined;
      return new TransformError(issue.message, { path: joined });
    });
    // Return the raw transformed object alongside errors so callers can
    // inspect malformed data.
    return { result: ran.value as TCommon, errors };
  };

  const fromCommon = (common: TCommon): TransformResult<TSource> => {
    const ran = runMapping(common, fromCommonMapping);
    if (!ran.ok) return { result: {} as TSource, errors: [ran.error] };

    if (sourceSchema === undefined) {
      return { result: ran.value as TSource, errors: [] };
    }

    const parsed = sourceSchema.safeParse(ran.value);
    if (parsed.success) {
      return { result: parsed.data, errors: [] };
    }
    const errors = parsed.error.issues.map(issue => {
      const joined = issue.path.length > 0 ? issue.path.map(p => String(p)).join(".") : undefined;
      return new TransformError(issue.message, { path: joined });
    });
    return { result: ran.value as TSource, errors };
  };

  return { toCommon, fromCommon };
}
